flink/src/main/java/org/apache/iceberg/flink/source/CombinedScanTaskIterator.java

```diff
@@ -0,0 +1,160 @@
+
+  private final FileIteratorReader<T> fileIteratorReader;
+  private final DecryptedInputFiles decryptedInputFiles;
+  private final Iterator<FileScanTask> tasks;
+  private final Position position;
+
+  private CloseableIterator<T> currentIterator;
+
+  public CombinedScanTaskIterator(CombinedScanTask combinedTask, FileIO io, EncryptionManager encryption,
+                                  FileIteratorReader<T> fileIteratorReader) {
+    this(combinedTask, io, encryption, fileIteratorReader, null);
+  }
+
+  public CombinedScanTaskIterator(CombinedScanTask combinedTask, FileIO io, EncryptionManager encryption,
+                                  FileIteratorReader<T> fileIteratorReader, @Nullable Position startingPosition) {
+    this.fileIteratorReader = fileIteratorReader;
+    this.decryptedInputFiles = new DecryptedInputFiles(combinedTask, io, encryption);
+    this.tasks = combinedTask.files().iterator();
+
+    if (startingPosition != null) {
+      this.position = startingPosition;
+      // skip files
+      Preconditions.checkArgument(position.fileOffset() < combinedTask.files().size(),
+          String.format("Starting file offset is %d, while CombinedScanTask has %d files",
+              position.fileOffset(), combinedTask.files().size()));
+      for (long i = 0L; i < position.fileOffset; ++i) {
+        tasks.next();
+      }
+    } else {
+      this.position = new Position();
+    }
+
+    final FileScanTask startingFileTask = tasks.next();
+    this.currentIterator = fileIteratorReader.open(startingFileTask, decryptedInputFiles);
+
+    // skip records
+    for (int i = 0; i < position.recordOffset(); ++i) {
+      if (currentIterator.hasNext()) {
+        currentIterator.next();
+      } else {
+        throw new IllegalArgumentException(String.format(
+            "File has less than %d records: %s", position.recordOffset, startingFileTask.file().path()));
+      }
+    }
+  }
+
+  @Override
+  public boolean hasNext() {
+    updateCurrentIterator();
+    return currentIterator.hasNext();
+  }
+
+  @Override
+  public T next() {
+    updateCurrentIterator();
+    position.advanceRecord();
+    return currentIterator.next();
+  }
+
+  public boolean isCurrentIteratorDone() {
+    return !currentIterator.hasNext();
+  }
+
+  /**
+   * Updates the current iterator field to ensure that the current Iterator is not exhausted.
+   */
+  private void updateCurrentIterator() {
+    try {
+      while (!currentIterator.hasNext() && tasks.hasNext()) {
+        currentIterator.close();
+        currentIterator = fileIteratorReader.open(tasks.next(), decryptedInputFiles);
+        position.advanceFile();
+      }
+    } catch (IOException e) {
+      throw new UncheckedIOException(e);
+    }
+  }
+
+  @Override
+  public void close() throws IOException {
+    // close the current iterator
+    currentIterator.close();
+  }
+
+  public Position position() {
+    return position;
+  }
+
+  public static class Position {
```

Comment:
@openinx following up your comment from the uber PR: https://github.com/apache/iceberg/pull/2105/files#r630834205.

The reason I introduce this mutable `Position` class is to avoid the construction of a <fileOffset, recordOffset> object. It is the current cursor for the iterator. 

Didn't track the recordOffset inside the `FileIteratorReader` for the same reason. Otherwise, `position()` getter will construct a new object each time.

We can't use CheckpointedPosition from Flink for two reasons: (1) it is immutable (2) we want to return the current position (not necessarily CheckpointedPosition).